### DISCUSSION TITLE
How to instrument a simple Go application

### DISCUSSION THREAD

**Discussion by Medik-GH** [keyval-dev/opentelemetry-go-instrumentation#16] [Q&A]
Hey,

Just wanted to say this is a really cool project. One thing I am confused about however is how to actually run this without the use of patch files and kubectl.

_Our goal is to provide the same level of automatic instrumentation for Go as exists for languages such as Java and Python._

For these languages (JavaScript also), the auto-instrumentation basically works by running some code with the agent acting as a sidecar.

If I have a simple main.go file that props up a web server and handles API requests and doesn't implement Kubernetes in any way, how can I auto-instrument that?

Can I do something as simple as importing this module or running my binary with some command line arguments?

Sorry if this is a basic question but I just didn't see anything in the docs.

**Comment by edeNFed** [comment]
Hi,
Happy to hear that you like the project!
If you are using Linux you can simply run the instrumentation of a process by doing the following:
1. Compile this project using `make build`
2. Run your process, for example `/usr/local/my-webserver`
3. Set the following env vars:
```
OTEL_TARGET_EXE=/usr/local/my-webserver
OTEL_EXPORTER_OTLP_ENDPOINT=<your-opentelemetry-endpoint>
OTEL_SERVICE_NAME=<your-app-name>
```
4. Run the binary created in step 1.

Please let me know how it went 😄

**Reply by Medik-GH** [reply]
Thanks, I'll give it a go :)

**Reply by Medik-GH** [reply]
@edeNFed 

```
opentelemetry-go-instrumentation on  master [!?] via 🐹 v1.18.2
❯ make build
go mod tidy
go generate ./...
Error: exec: "llvm-strip": executable file not found in $PATH
exit status 1
pkg/instrumentors/bpf/github.com/gorilla/mux/probe.go:22: running "go": exit status 1
Error: exec: "llvm-strip": executable file not found in $PATH
exit status 1
pkg/instrumentors/bpf/google/golang/org/grpc/probe.go:23: running "go": exit status 1
Error: exec: "llvm-strip": executable file not found in $PATH
exit status 1
pkg/instrumentors/bpf/google/golang/org/grpc/server/probe.go:22: running "go": exit status 1
Error: exec: "llvm-strip": executable file not found in $PATH
exit status 1
pkg/instrumentors/bpf/net/http/server/probe.go:22: running "go": exit status 1
make: *** [generate] Error 1
```
```
❯ echo $PATH
/usr/local/opt/llvm/bin/llvm-strip 
```

Any ideas? I'm on a Mac. I installed llvm with ```brew install llvm``` and then added llvm-strip to the $PATH but still no dice.

Alternatively, if you could upload the compiled binary is an artifact somewhere that would be awesome :D

**Reply by Medik-GH** [reply]
Well I got the above issue sorted but now I get
```
/Users/eeldridge/Documents/Personal/opentelemetry-go-instrumentation/pkg/instrumentors/bpf/headers/arguments.h:2:10: fatal error: 'bpf_helpers.h' file not found
#include "bpf_helpers.h"
Error: clang: exit status 1
exit status 1
pkg/instrumentors/bpf/net/http/server/probe.go:22: running "go": exit status 1
make: *** [generate] Error 1
```

I don't think it's going to be possible to compile this on Mac.

**Comment by Medik-GH** [comment]
Ok @edeNFed, everything is looking good! I compiled it with the Docker command provided and just copied out the compiled binary from the container.

I just now have this problem

```json
{
    "level": "error",
    "ts": 1653408162.9497857,
    "caller": "opentelemetry/controller.go:101",
    "msg": "unable to connect to OpenTelemetry collector",
    "addr": "http://localhost:4318",
    "error": "context deadline exceeded",
    "stacktrace": "github.com/keyval-dev/opentelemetry-go-instrumentation/pkg/opentelemetry.NewController\n\t/app/pkg/opentelemetry/controller.go:101\nmain.main\n\t/app/cli/main.go:30\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"
}
{
    "level": "error",
    "ts": 1653408162.9499285,
    "caller": "cli/main.go:32",
    "msg": "unable to create OpenTelemetry controller",
    "error": "context deadline exceeded",
    "stacktrace": "main.main\n\t/app/cli/main.go:32\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"
}
```

I had a quick Google and it seems like perhaps this is some kind of security issue? Is it possible to tell the binary where to look for the patch file? That would allow me to configure the security context like here

```yaml
securityContext:
    runAsUser: 0
    capabilities:
      add:
        - SYS_PTRACE
    privileged: true
```

**Reply by edeNFed** [reply]
Hi @EddieEldridge it looks like the instrumentation agent can't communicate with the address configured at `OTEL_EXPORTER_OTLP_ENDPOINT` and getting a timeout.
Does making requests to `http://localhost:4318` work? 
In addition i don't think you need the `http://` it might also be a problem, try set the env var to `localhost:4318`

**Reply by Medik-GH** [reply]
@edeNFed Thanks, I actually needed to use ```locahost:4317``` which is the default endpoint for gRPC connections (https://github.com/open-telemetry/opentelemetry-collector/blob/main/examples/k8s/otel-config.yaml#L142). Previously we were only using HTTP, hence my choice of 4318.

Thanks for helping me out as it's now connecting fine.

Now we have another interesting error however,

```json
{
    "level": "error",
    "ts": 1653475250.2704098,
    "caller": "instrumentors/runner.go:64",
    "msg": "error loading goroutine tracker",
    "error": "field UprobeRuntimeCasgstatus: program uprobe_runtime_casgstatus: map .rodata: map create: read- and write-only maps not supported (requires >= v5.2)",
    "stacktrace": "github.com/keyval-dev/opentelemetry-go-instrumentation/pkg/instrumentors.(*instrumentorsManager).load\n\t/app/pkg/instrumentors/runner.go:64\ngithub.com/keyval-dev/opentelemetry-go-instrumentation/pkg/instrumentors.(*instrumentorsManager).Run\n\t/app/pkg/instrumentors/runner.go:19\nmain.main\n\t/app/cli/main.go:71\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"
}
{
    "level": "error",
    "ts": 1653475250.2704575,
    "caller": "cli/main.go:73",
    "msg": "error while running instrumentors",
    "error": "field UprobeRuntimeCasgstatus: program uprobe_runtime_casgstatus: map .rodata: map create: read- and write-only maps not supported (requires >= v5.2)",
    "stacktrace": "main.main\n\t/app/cli/main.go:73\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"
}
```

I had a read through https://github.com/cilium/ebpf/issues/628 and it seems to indicate that perhaps this possibly only supports Linux Kernel versions > 5 ?


The kernel version on my instance is 
```
[eddie.eldridge@exampleservice-go-********* ~]$ uname -srm
Linux 4.14.276-211.499.amzn2.x86_64 x86_64
```

Any ideas?

**Reply by elskndd** [reply]
hi, @edeNFed @EddieEldridge 
When I tried this, I found it can connect to connector with port 4317, but cannot with port 4318. If using port 4318, will get the "context deadline exceeded" error.

`OTEL_TARGET_EXE=/home/elskndd/project/httptest/main OTEL_EXPORTER_OTLP_ENDPOINT=localhost:4318 OTEL_SERVICE_NAME=my-httptest ./kv-go-instrumentation`
{"level":"info","ts":1670374839.2088923,"caller":"cli/main.go:22","msg":"starting Go OpenTelemetry Agent ..."}
{"level":"info","ts":1670374839.2093885,"caller":"opentelemetry/controller.go:92","msg":"Establishing connection to OpenTelemetry collector ..."}
{"level":"error","ts":1670374849.2123194,"caller":"opentelemetry/controller.go:97","msg":"unable to connect to OpenTelemetry collector","addr":"localhost:4318","error":"context deadline exceeded","stacktrace":"github.com/keyval-dev/opentelemetry-go-instrumentation/pkg/opentelemetry.NewController\n\t/home/elskndd/project/opentelemetry-go-instrumentation/pkg/opentelemetry/controller.go:97\nmain.main\n\t/home/elskndd/project/opentelemetry-go-instrumentation/cli/main.go:30\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"}
{"level":"error","ts":1670374849.2125142,"caller":"cli/main.go:32","msg":"unable to create OpenTelemetry controller","error":"context deadline exceeded","stacktrace":"main.main\n\t/home/elskndd/project/opentelemetry-go-instrumentation/cli/main.go:32\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"}

I tried `curl` with the endpoint. The collector is working.
`curl http://localhost:4318/v1/metrics`
`405 method not allowed, supported: [POST]`

The collector is a docker container
`docker run -v $(pwd)/collector-config-test.yaml:/etc/otelcol/config.yaml -p 4318:4318 otel/opentelemetry-collector:0.66.0`
The configuration of collector is as following.

receivers:
  otlp:
    protocols:
      grpc:
      http:

processors:
  memory_limiter:
    check_interval: 5s
    limit_mib: 4000
    spike_limit_mib: 500
  batch:

exporters:
  logging:
    loglevel: debug

service:
  pipelines:
    metrics:
      receivers: [otlp]
      exporters: [logging]
    traces:
      receivers: [otlp]
      processors: [batch]
      exporters: [logging]



I also tried localhost:4317, but got another error after the test application ran.

`OTEL_TARGET_EXE=/home/elskndd/project/httptest/main OTEL_EXPORTER_OTLP_ENDPOINT=localhost:4317 OTEL_SERVICE_NAME=my-httptest ./kv-go-instrumentation`
{"level":"info","ts":1670375178.49957,"caller":"cli/main.go:22","msg":"starting Go OpenTelemetry Agent ..."}
{"level":"info","ts":1670375178.5002918,"caller":"opentelemetry/controller.go:92","msg":"Establishing connection to OpenTelemetry collector ..."}
{"level":"info","ts":1670375180.529661,"caller":"process/discover.go:37","msg":"process not found yet, trying again soon","exe_path":"/home/elskndd/project/httptest/main"}
{"level":"info","ts":1670375182.5116515,"caller":"process/discover.go:37","msg":"process not found yet, trying again soon","exe_path":"/home/elskndd/project/httptest/main"}
{"level":"info","ts":1670375184.5074406,"caller":"process/discover.go:37","msg":"process not found yet, trying again soon","exe_path":"/home/elskndd/project/httptest/main"}
{"level":"info","ts":1670375186.510753,"caller":"process/discover.go:37","msg":"process not found yet, trying again soon","exe_path":"/home/elskndd/project/httptest/main"}
{"level":"info","ts":1670375188.5061173,"caller":"process/discover.go:37","msg":"process not found yet, trying again soon","exe_path":"/home/elskndd/project/httptest/main"}
{"level":"info","ts":1670375190.5098116,"caller":"process/discover.go:37","msg":"process not found yet, trying again soon","exe_path":"/home/elskndd/project/httptest/main"}
{"level":"info","ts":1670375192.505927,"caller":"process/discover.go:42","msg":"found process","pid":2865}
panic: cant find keyval map

goroutine 1 [running]:
github.com/keyval-dev/opentelemetry-go-instrumentation/pkg/process.(*processAnalyzer).findKeyvalMmap(0xb70120?, 0xc00011d6a8?)
        /home/elskndd/project/opentelemetry-go-instrumentation/pkg/process/analyze.go:77 +0x14f
github.com/keyval-dev/opentelemetry-go-instrumentation/pkg/process.(*processAnalyzer).Analyze(0xc000088150?, 0xb31, 0x2?)
        /home/elskndd/project/opentelemetry-go-instrumentation/pkg/process/analyze.go:103 +0x1a9
main.main()
        /home/elskndd/project/opentelemetry-go-instrumentation/cli/main.go:59 +0x37f.


Have any ideas?

**Reply by elskndd** [reply]
Found it. From the code, it seems just supports grpc.
https://github.com/keyval-dev/opentelemetry-go-instrumentation/blob/483a2658fbf4d58ae0ef997709c0fac44966bf99/pkg/opentelemetry/controller.go#L95

**Comment by Medik-GH** [comment]
@edeNFed Should I close this and make a PR to outline the required Kernel version?

**Reply by edeNFed** [reply]
Interesting. Let me try to find a fix that will work on older kernels. Will update soon

**Reply by mark800** [reply]
@edeNFed Thanks for contribute this cool project. 

I wrote simple a gRpc Server and client program (the client send a string to the server, and the server just reply a hello). It seems the “kv-go-instrumentation”can’t output trace info automatically.
Every time the client sends a string to the Server, got a log like this:

{"level":"info","ts":1663581820.4961782,"caller":"opentelemetry/controller.go:44","msg":"got event","attrs":[{"Key":"rpc.system","Value":{"Type":"STRING","Value":"grpc"}},{"Key":"rpc.service","Value":{"Type":"STRING","Value":""}}],"goroutine":0}

And there is no trace in the Jaeger UI. (Can find the service name in Jaeger)
